framework/test/gtest/BundleTest.cpp

```diff
@@ -0,0 +1,906 @@
+                         "TestBundleExt" + US_LIB_POSTFIX + ".cppms");
+
+  // 3 bundles - the framework(system_bundle), the executable(main) and TextBundleExt
+  ASSERT_EQ(3, context.GetBundles().size());
+#else
+  // There are atleast 3 bundles, maybe more depending on how the executable is created
+  EXPECT_GE(static_cast<int>(context.GetBundles().size()));
+#endif
+
+  // Test the non-standard file extension bundle's lifecycle
+  auto bundles = context.GetBundles();
+  auto bundleIter =
+    std::find_if(bundles.begin(), bundles.end(), [](const Bundle& b) {
+      return (std::string("TestBundleExt") == b.GetSymbolicName());
+    });
+
+  ASSERT_NE(bundleIter, bundles.end());
+  (*bundleIter).Start();
+  (*bundleIter).Uninstall();
+}
+
+TEST_F(BundleTest, TestUnicodePaths)
+{
+// 1. building static libraries (test bundle is included in the executable)
+// 2. using MINGW evironment (MinGW linker fails to link DLL with unicode path)
+// 3. using a compiler with no support for C++11 unicode string literals
+#if defined(US_BUILD_SHARED_LIBS) || !defined(__MINGW32__) ||                  \
+  defined(US_CXX_UNICODE_LITERALS)
+  std::string path_utf8 = LIB_PATH + cppmicroservices::util::DIR_SEP +
+                          u8"くいりのまちとこしくそ" +
+                          cppmicroservices::util::DIR_SEP + US_LIB_PREFIX +
+                          "TestBundleU" + US_LIB_POSTFIX + US_LIB_EXT;
+  auto bundles = context.InstallBundles(path_utf8);
+  ASSERT_EQ(bundles.size(), 1);
+  auto bundle = bundles.at(0);
+  //Bundle location is the same as the path used to install
+  ASSERT_EQ(bundle.GetLocation(), path_utf8);
+  bundle.Start();
+  //Bundle check start state
+  ASSERT_EQ(bundle.GetState(), Bundle::State::STATE_ACTIVE);
+  bundle.Stop();
+
+#endif
+}
+
+TEST_F(BundleTest, TestBundleStartOptions)
+{
+  auto bundle = InstallLib(context, "TestBundleA");
+  EXPECT_NO_THROW(bundle.Start(
+    cppmicroservices::Bundle::StartOptions::START_ACTIVATION_POLICY));
+}
+
+TEST_F(BundleTest, TestBundleLessThanOperator)
+{
+
+  auto bundleA = InstallLib(context, "TestBundleA");
+  auto bundleB = InstallLib(context, "TestBundleB");
+
+  //Test that bundles are ordered correctly.
+  ASSERT_TRUE(bundleA < bundleB);
+  ASSERT_TRUE(bundleA < Bundle());
+  ASSERT_TRUE(!(Bundle() < bundleB));
+  ASSERT_TRUE(!(Bundle() < Bundle()));
+}
+
+TEST_F(BundleTest, TestBundleAssignmentOperator)
+{
+  Bundle b;
+  Bundle b1;
+  b = b1;
+  //Test that the bundle objects are equivalent.
+  ASSERT_EQ(b1, b);
+
+  auto bundleA = InstallLib(context, "TestBundleA");
+  auto bundleB = InstallLib(context, "TestBundleB");
+
+  //Test that the bundles are different before assignment.
+  ASSERT_NE(bundleB.GetBundleId(), bundleA.GetBundleId());
+  ASSERT_NE(bundleB, bundleA);
+
+  bundleB = bundleA;
+
+  //Test bundle object assignment.
+  ASSERT_EQ(bundleB.GetBundleId(), bundleA.GetBundleId());
+  //Test that the bundle objects are equivalent.
+  ASSERT_EQ(bundleB, bundleA);
+}
+
```

Comment:
missing a test, it is `TestBundleStreamOperator()` in the test driver

Reply:
TestBundleStreamOperator() simply installs 'TestBundleA' (which is being tested in other testpoints) and gives out bundle information (id, location, etc.) as an output.
There was not any validations as such. Hence, I removed it.

Reply:
The operation of outputting the `Bundle` object is what is being tested. Unfortunately there wasn't a testpoint for it that validated a behavior. However running that code can uncover bugs in the `Bundle`'s operator<<.

Reply:
Ill add it to the test